File: webapp/src/main/scala/org/linnando/mixemulator/webapp/MemoryTextComponent.scala
package org.linnando.mixemulator.webapp

import angulate2.core.{AfterViewInit, ElementRef}
import angulate2.std._

import scala.scalajs.js

@Component(
  selector = "memory-text",
  templateUrl = "webapp/src/main/resources/memory-text.component.html",
  styleUrls = @@@("webapp/src/main/resources/memory-text.component.css")
)
class MemoryTextComponent(virtualMachineService: VirtualMachineService) extends OnInit with AfterViewInit {
  @ViewChild("scrolling")
  var scrolling: ElementRef = _

  override def ngOnInit(): Unit = {
    virtualMachineService.stateChange.subscribe(_ => positionToProgramCounter())
  }

  override def ngAfterViewInit(): Unit = {
    positionToProgramCounter()
  }

  private def positionToProgramCounter(): Unit = virtualMachineService.programCounterIndex match {
    case Some(index) =>
      val scrollHeight = scrolling.nativeElement.scrollHeight.asInstanceOf[Double]
      val clientHeight = scrolling.nativeElement.clientHeight.asInstanceOf[Double]
      scrolling.nativeElement.scrollTop = index * scrollHeight / virtualMachineService.symbolsLength - clientHeight / 2
    case None =>
  }

  def indices: js.Array[Int] = js.Array[Int]() ++ virtualMachineService.symbolIndices

  def toggleBreakpoint(index: Int): Unit = virtualMachineService.toggleBreakpointAt(index)

  def hasBreakpointAt(index: Int): Boolean = virtualMachineService.breakpointAt(index).exists(b => b)

  def isCurrent(index: Int): Boolean = virtualMachineService.programCounterIndex.contains(index)

  def hasWordAt(index: Int): Boolean = {
    val address = virtualMachineService.addressAt(index)
    address.isDefined
  }

  def cellAddressAt(index: Int): Short = {
    val address = virtualMachineService.addressAt(index)
    address.getOrElse(0)
  }

  def cellSignAt(index: Int): String = {
    val cellContent = virtualMachineService.cellContent(index)
    cellContent.map(w => if (w.negative) "-" else "+").getOrElse("+")
  }

  def cellByteAt(index: Int, pos: Int): Byte = {
    val cellContent = virtualMachineService.cellContent(index)
    cellContent.map(_.bytes(pos)).getOrElse(0)
  }

  def hasLineAt(index: Int): Boolean = {
    val lineNumber = virtualMachineService.lineNumberAt(index)
    lineNumber.isDefined
  }

  def lineNumberAt(index: Int): Int = {
    val lineNumber = virtualMachineService.lineNumberAt(index)
    lineNumber.getOrElse(0)
  }

  def lineAt(index: Int): String = {
    val line = virtualMachineService.lineAt(index)
    line.getOrElse("")
  }

  def lineIsModifiedAt(index: Int): Boolean = virtualMachineService.lineIsModifiedAt(index).getOrElse(false)
}

File: webapp/src/app/memory-text/memory-text.component.ts
import { AfterViewInit, Component, ElementRef, OnDestroy, OnInit, ViewChild } from '@angular/core';
import { IOWord, IOWord$ } from '@mixemulator/lib';
import { Subscription } from 'rxjs';
import { VirtualMachineService } from '../virtual-machine.service';

@Component({
  selector: 'mix-memory-text',
  templateUrl: './memory-text.component.html',
  styleUrls: ['./memory-text.component.scss']
})
export class MemoryTextComponent implements OnInit, AfterViewInit, OnDestroy {
  @ViewChild("scrolling") scrolling!: ElementRef;
  stateChangeSubscription?: Subscription;

  constructor(private virtualMachineService: VirtualMachineService) { }

  ngOnInit(): void {
    this.stateChangeSubscription = this.virtualMachineService.stateChange.subscribe(_ => {
      this.positionToProgramCounter();
    });
  }

  ngAfterViewInit(): void {
    this.positionToProgramCounter();
  }

  private positionToProgramCounter(): void {
    if (this.virtualMachineService.programCounterIndex != null) {
      const scrollHeight = this.scrolling.nativeElement.scrollHeight;
      const clientHeight = this.scrolling.nativeElement.clientHeight;
      this.scrolling.nativeElement.scrollTop = this.virtualMachineService.programCounterIndex * scrollHeight / this.virtualMachineService.symbolsLength - clientHeight / 2;
    }
  }

  ngOnDestroy(): void {
    this.stateChangeSubscription?.unsubscribe();
  }

  get indices(): number[] {
    return this.virtualMachineService.symbolIndices;
  }

  toggleBreakpoint(index: number): void {
    this.virtualMachineService.toggleBreakpointAt(index);
  }

  hasBreakpointAt(index: number): boolean {
    return this.virtualMachineService.breakpointAt(index) ?? false;
  }

  isCurrent(index: number): boolean {
    return this.virtualMachineService.programCounterIndex === index;
  }

  hasWordAt(index: number): boolean {
    return this.virtualMachineService.addressAt(index) != null;
  }

  cellAddressAt(index: number): number {
    return this.virtualMachineService.addressAt(index) ?? 0;
  }

  cellSignAt(index: number): string {
    return this.cellContentAt(index).negative ? '-' : '+';
  }

  private cellContentAt(index: number): IOWord {
    return this.virtualMachineService.cellContent(index) ?? IOWord$.empty();
  }

  cellByteAt(index: number, pos: number): number {
    return this.cellContentAt(index).bytes[pos];
  }

  hasLineAt(index: number): boolean {
    return this.virtualMachineService.lineNumberAt(index) != null;
  }

  lineNumberAt(index: number): number {
    return this.virtualMachineService.lineNumberAt(index) ?? 0;
  }

  lineAt(index: number): string {
    return this.virtualMachineService.lineAt(index) ?? '';
  }

  lineIsModifiedAt(index: number): boolean {
    return this.virtualMachineService.lineIsModifiedAt(index) ?? false;
  }
}
